### DISCUSSION TITLE
is there a `state.next` method for `SpinalEnum?

### DISCUSSION THREAD

**Discussion by goekce** [SpinalHDL/SpinalHDL#1599] [Q&A] [answered]
SystemVerilog has the `next` method, which I find quite readable. Does SpinalHDL have something similar?

I could not find something interesting in https://github.com/SpinalHDL/SpinalHDL/blob/dev/core/src/main/scala/spinal/core/Enum.scala

**Comment by Dolu1990** [comment, marked as answer]
Added ^^

**Comment by Dolu1990** [comment]
(https://github.com/SpinalHDL/SpinalHDL/commit/09320225c161c84b4ff83d40641fb2b3b92da13f)

**Reply by goekce** [reply]
Thanks @Dolu1990  🙂

Note that in SystemVerilog `enum.next`:
1.  a wrap to the start of the enumeration occurs when the end of the enumeration is reached, which I find also very useful in case of fully sequential FSMs with almost no branches.
     In the current impl you return null in case of wrap, which I find more conservative and acceptable. 
      I would probably go for wrapping, because conservative programmers probably avoid prev and next.
4. the signature in SystemVerilog is `next(unsigned int N=1)`, which allows jumps which I find also semantically useful.

These apply to `prev` too. LRM 6.19.5

**Comment by Dolu1990** [comment]
what's about having : 
- next 
- nextWrapped 

?

**Reply by goekce** [reply]
This is a good trade-off!
